docs: Add document for environment variables

terrytangyuan opened the pull request:
Currently we have the use of environment variables in many places at different levels but there's no documentation on them. This PR takes an initial stab at documenting them. 

cc @alexec @simster7 

Signed-off-by: terrytangyuan <terrytangyuan@gmail.com>

Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

<!--

Please leave your PR in draft if you don't need a review yet. 

To fix failing `CI / Codegen`, run `make codegen`. 

-->

simster7 reviewed (commented):
It'd be awesome if this could be auto-generated by looking at source files. This would also force the documentation to appear in someway in the location where it is used. Do you think you could manage a script to do that @terrytangyuan ?

simster7 commented:
Similar to how all of our other docs are generated (`fields.md` from `*_types.go`, etc.)

alexec commented on docs/environment-variables.md:
```diff
@@ -0,0 +1,46 @@
+# Environment Variables
+
+This document outlines the set of environment variables that can be set to customize the behaviours at different levels.
```
It needs to be clearly stated that these are unsupported and may be removed.

terrytangyuan replied:
Good point. Just added a note on this.

terrytangyuan commented:
> It'd be awesome if this could be auto-generated by looking at source files. This would also force the documentation to appear in someway in the location where it is used. Do you think you could manage a script to do that @terrytangyuan ?

@simster7 Yes that's definitely achievable. However I think it's a bit overkill and it doesn't seem to require a lot of efforts maintaining this document at this point. I agree that in the long term it makes sense to automate so it's easy to maintain but perhaps not a priority for now.

alexec commented on docs/environment-variables.md:
```diff
@@ -0,0 +1,47 @@
+| `REMOVE_LOCAL_ART_PATH` | bool | Whether to remove local artifacts. |
+
+
+## CLI
```
I think can exclude CLI ones, you can run `argo --help`

terrytangyuan replied:
Sure. Removed this section.

alexec commented on docs/environment-variables.md:
```diff
@@ -0,0 +1,36 @@
+# Environment Variables
+
+This document outlines the set of environment variables that can be used to customize the behaviours at different levels.
```
I think I'm wrong about the text for this. I think we should not emphasize they are unsupported, more that they are typically added to test out experimental features and should not be needed by most users.

terrytangyuan replied:
That sounds much better actually. I just pasted the exact sentence you said but feel free to modify further.

simster7 replied:
I would actually explicitly say that these may be removed at any time

alexec commented:
@simster7 before merge, you always have valuable opinions on docs, do you want to take a look?

simster7 commented on docs/environment-variables.md:
```diff
@@ -0,0 +1,35 @@
+|----------|------|------------|
+| `ALL_POD_CHANGES_SIGNIFICANT` | bool |  Whether to consider all pod changes as significant during pod reconciliation. |
+| `ALWAYS_OFFLOAD_NODE_STATUS` | bool | Whether to always offload the node status. |
+| `ARCHIVED_WORKFLOW_GC_PERIOD` | time.duration | The periodicity for GC of archived workflows. |
```
Very minor, but correct capitalization emphasizes that this is the Go type

```suggestion
| `ARCHIVED_WORKFLOW_GC_PERIOD` | time.Duration | The periodicity for GC of archived workflows. |
```

simster7 commented on docs/environment-variables.md:
```diff
@@ -0,0 +1,35 @@
+
+| Name | Type | Description|
+|----------|------|------------|
+| `ALL_POD_CHANGES_SIGNIFICANT` | bool |  Whether to consider all pod changes as significant during pod reconciliation. |
```
I would also wrap types in in-line codeblocks to be consistent with our docs elsewhere

```suggestion
| `ALL_POD_CHANGES_SIGNIFICANT` | `bool` |  Whether to consider all pod changes as significant during pod reconciliation. |
```

simster7 reviewed (changes requested):
I still believe that this should be auto generated, or at the very least it should have an automatic check that all of the flags here still exist in the codebase and that no flags that exist in the code base are not here. Otherwise this file would in practice get neglected and become out of date soon.

This could be done with a simple regex:
<img width="1241" alt="image" src="https://user-images.githubusercontent.com/614838/107666798-c04fdc80-6c43-11eb-8851-85a7d454cbaa.png">

terrytangyuan commented:
@simster7 Thanks! Just addressed the comments.

I probably don't have width to work on automating this at this point yet. Anyone else please feel free to contribute that before I get to it. Also note that besides `env.LookupEnv**` from our own utils (which we only have a few that uses those), there are also uses of `os.LookupEnv`, `os.Getenv`, etc. that should be taken into account when writing the automation code.

simster7 commented on docs/environment-variables.md:
```diff
@@ -0,0 +1,36 @@
+| `LEADER_ELECTION_IDENTITY` | `string` | The ID used for workflow controllers to elect a leader. |
+| `MAX_OPERATION_TIME` | `time.Duration` | The maximum time a workflow operation is allowed to run for before requeuing the workflow onto the work queue. |
+| `OFFLOAD_NODE_STATUS_TTL` | `time.Duration` | The TTL to delete the offloaded node status. Currently only used for testing. |
+| `POD_GC_TRANSIENT_ERROR_MAX_REQUEUES` | `int` | The maximum number of requeues of pods in the GC queue when encountering transient errors. |
```
Actually, can you remove this and add it in https://github.com/argoproj/argo-workflows/pull/5060 so that it doesn't fail the check when I merge https://github.com/argoproj/argo-workflows/pull/5091?

terrytangyuan replied:
Removed and rebased.

terrytangyuan replied:
I'll add this in #5060 once this has been merged.
